### ISSUE TITLE
Puppeteer script fails, fetch as google fails

### ISSUE THREAD

**Issue by otterslide** [malchata/yall.js#52]
I implemented Yall, and it loads very nicely on my desktop, and even the Puppeteer screenshot, but the script fails because the comparison is different, and Fetch as google shows a blank.

This is what Google says:
https://developers.google.com/search/docs/guides/lazy-loading

Test
After you set up your implementation, you should make sure it works correctly. One way you do this is by using a Puppeteer script to locally test your implementation. Puppeteer is a Node.js library for controlling headless Chrome. To run the script, you'll need Node.js. Use the following commands to check out the script and run it:

So without scroll events, it fails.. I don't know what reason? It is using intersection observer, but still does not work through this script. Worse is that Google fetch fails completely...

I even tried to Polyfill, and it doesn't load in fetch as Google. Very strange.

Any ideas? Have you tested this on Fetch as Google and Puppeteer?
Only Zepto Lazy was able to load properly with Fetch as Google.  Even the famous verlok/lazyload did not work, but the old Fetch as Google is gone now. It would load only the first image. Now we can never know if the fetch works properly for 2nd, 3rd images, etc.. only first.

Just to note, the Puppeteer screenshot with scroll does end up loading the images. But the one without scroll is missing them.

I have yet to find one script that passes that test, and most other don't pass the fetch as Google test. 

Seems like this lazy loading is more of a headache than anything concerning compatibility and Google. I've spent so much time trying to find a good solution.

**Comment by malchata** [comment]
Even though the old Fetch as Google tool is gone, their new search console offers the same functionality by way of the URL inspector. To get there, search for a URL in the index for your site > click URL inspection on the left hand bar > view tested page > screenshot. You may have to tell the tool to fetch a screen grab, but once one is available, you can scroll down in the preview pane to see the full screenshot. I can confirm yall.js is exposing lazy loaded images to Google, at least for my site.

![image](https://user-images.githubusercontent.com/1635238/55671298-69fa0180-5854-11e9-9372-3f42c35fbe99.png)

Although, Googlebot is _not_ triggering the events I've set on those images, which is why the lazy-loaded images are transparent, but those are the final images. I can also see that Google is indexing at least _some_ of those images when I search for ["jeremy wagner developer" in their image search](https://www.google.com/search?q=jeremy+wagner+developer&source=lnms&tbm=isch&sa=X&ved=0ahUKEwiB6J6R4LvhAhUN24MKHbmbAbEQ_AUIDygC&biw=1280&bih=649&dpr=2).

As for why the scrolling test fails on Puppeteer, that's a simple answer: I don't perform a check like `"onscroll" in window`. Maybe I ought to do that, but on the other hand, my images are being indexed in Google, but I can't find them in other search engines.

If your site has been up for a while with yall.js 3.x, I'd say do an image search and see if your lazy-loaded images are showing up in the index. If they are, it's apparently working. If not, there may be something a bit more nuanced afoot. Are you using yall.js live in prod now? If so, do you have a URL I can take a look at?

**Comment by otterslide** [comment]
The new Fetch tool image is only 732px high. I can't see the full website screenshot like before.
Did you get that screenshot from Google? 
I made a test URL here using Yall:
https://www.purelovequotes.com/test.html

I even put the Yall script right in the head tag, but the Google fetch tool still doesn't show an image.

![Screenshot from 2019-04-06 11-47-46](https://user-images.githubusercontent.com/11649943/55671810-e7783e80-5861-11e9-8e81-9e91d8469d42.png)

Thanks for the help.

**Comment by otterslide** [comment]
The problem I had with other Lazy Loading scripts is that the Google DFP/Ad tags were taking precedence, and images don't load until ads are loaded.

I have no idea why your approach is somehow working in parallel as the ad loading tags. Perhaps because intersection observer is supposed to work in another thread.
In your article:
https://developers.google.com/web/fundamentals/performance/lazy-loading-guidance/images-and-video/
you recommend a few. But Lazysizes for example doesn't use Intersection observer and has the ad loading problem. I also don't understand how the ad loading can block the thread so efficiently as to block all scrolling events, I didn't think this should happen, but it happens on my site.  The .ready() is called, script is initialized, but never loads any images. Your script is the only one so far that works in parallel and I tried about 4 of them.

**Comment by otterslide** [comment]
Could this be he issue: I have a default hidden class on lazy images from the previous script I was using:
.lazy {
    visibility: hidden;
}
Your script seems to set the visibility to remove the "lazy" class. I don't like the broken image icon if not using the visibility tag. On your site, do you see the broken image tag with the Alt text before the image loads?

**Comment by malchata** [comment]
Sorry, I haven't been able to catch up to this entire thread yet, but yall.js does not set any styles, only classes, which you may or may not decide to add styles to. I don't see issues with broken image tags on my site because I'm using placeholder images. Check out [the README](https://github.com/malchata/yall.js/blob/master/README.md) for more info.

**Comment by otterslide** [comment]
> Sorry, I haven't been able to catch up to this entire thread yet, but yall.js does not set any styles, only classes, which you may or may not decide to add styles to. I don't see issues with broken image tags on my site because I'm using placeholder images. Check out [the README](https://github.com/malchata/yall.js/blob/master/README.md) for more info.

I tried removing the style, but Fetch as Google still does not show.  I just tried Lozad and it is working on fetch as Google and using Intersection Observer. I may just use that for now.

It would be great if you could confirm if fetch as google is giving you a bigger screenshot. It would help testing these scripts. So far, I cannot get more than 732px height. Before it was 10,000px... 

Thanks.

**Comment by malchata** [comment]
No worries. If another solution does the trick for you, definitely use it! The screenshot of my site I posted above was from Fetch as Google. You may want to download a copy of Chrome 41 to see if your site runs in it, as that's the version of Chrome Googlebot runs. I was struggling with fixing SEO in yall.js, but didn't know that my bundler was not transpiling code properly to run in legacy browsers, and that—at least in part—was causing Chrome to not lazy load images.

**Comment by otterslide** [comment]
> No worries. If another solution does the trick for you, definitely use it! The screenshot of my site I posted above was from Fetch as Google. You may want to download a copy of Chrome 41 to see if your site runs in it, as that's the version of Chrome Googlebot runs. I was struggling with fixing SEO in yall.js, but didn't know that my bundler was not transpiling code properly to run in legacy browsers, and that—at least in part—was causing Chrome to not lazy load images.

That's fetch as Google done with the new tool?
Could you look at the screenshot I posted above of the fetch as Google? Am I in the correct place?

All I see is what is in the screenshot, I cannot get any bigger image.
I removed all JS except for your script on the test URL. Fetched as Google, and it still shows blank for the image. There is no bundling or anything, not even external JS.
And it's working with my browser. I will get Chrome 41 and see what happens.

Thanks for your help! I really hope I can get that longer screenshot, but at this point I don't see how?

**Comment by otterslide** [comment]
I used Chromium 41 and it's working fine.
Googlebot must be using something different somehow, because the test url shows no image in the Fetch as google, but it does show an image with Chromium.
Also, at the bottom it says "Rendered with Googlebot smartphone".
Are you sure the smartphone version is still using Chrome 41?

**Comment by otterslide** [comment]
Also I found that the Google Mobile Friendly tester shows the same rendered image as the Fetch As Google. Here is the result using Yall:
https://search.google.com/test/mobile-friendly?id=aPMQ6W5XoQh-RO6WC84xsw

**Comment by otterslide** [comment]
Your site is failing as well:
https://search.google.com/test/mobile-friendly?id=fcgWHLV1pa87FsijMojdpQ

**Comment by malchata** [comment]
https://search.google.com/test/mobile-friendly?id=8sne7i7X_6FTqe3VFQVWFA

:shrug:

As for Googlebot, I can confirm it's using Chrome 41, at least as of a
couple weeks ago, because in the fetch as google tool, I wrote some JS to
dump the user agent string in the document, and it indicates version 41.

**Comment by otterslide** [comment]
Well, something is strange for sure. The old Fetch tool seems to be using a
different rendering algorithm than the new tool, because you were able to
render that desktop view and it worked fine.
So either the new tool is more like actual Google, or the new tool is using
some other kind of rendering and not 100% based on how Google sees the site.
My guess is that the new tool is incorrectly rendering, and it's very
annoying that Google has chosen to give us this inadequate tool when the
old one was working fine.

**Comment by malchata** [comment]
Since Googlebot's version of Chrome is now evergreen, I'm not concerned that this is necessarily an issue anymore. Closing. If you still have issues, create a new ticket and document it there.
